### PR TITLE
Implementing TransformerClassifier 

### PR DESCRIPTION
**Pull request by SalmanMohammadi**
#### Context/Changelog
See https://github.com/pytorch/torchtune/issues/812

This PR adds a `TransformerClassifier` layer which extends the `TransformerDecoder` functionality to classification tasks. Exemplar component and model builders have been implemented for the base `mistral` model.

#### Test plan

Testing this was tricky as there is currently no reference implementation for base `mistral` to test against. I performed some numerical testing against HuggingFace `MistralModel` and `AutoModelForSequenceClassification` with `mistralai/Mistral-7B-v0.1`, with the same parameters found in [`test_llama3.py`](https://github.com/pytorch/torchtune/blob/main/tests/torchtune/models/test_llama3.py). However, neither the classification model outputs, nor the base `MistralModel` and `torchtune.models.mistral.mistral` produced similar outputs to the HF models.
I've left a sort-of dummy test in which just asserts the output shapes are correct.

### Questions/Next steps

This is part of a broader plan to implement RLHF in Torchtune. The `TransformerClassifier` can hopefully be used against any sequence model we have. @kartikayk - we could implement a recipe for training a mistral reward model using this. I can start implementing a PPO recipe using this reward model, too.

I don't currently have access to sufficient compute to test converting a full 7b pre-trained mistral reward model, but I'm looking into this.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -39,7 +39,7 @@ tune = "torchtune._cli.tune:main"
 
 [project.optional-dependencies]
 dev = [
-    "bitsandbytes>=0.43.0",
+    # "bitsandbytes>=0.43.0",
```
[the remark this answers is not in the document]
Comment (SalmanMohammadi) [reply pyproject.toml:42]: Yes :) and failing pre-commits because I was pushing to `main` on my fork. I'll fix this and put a new PR up.
